src/sync/channel.zig

```diff
@@ -128,19 +143,27 @@ pub fn Channel(T: type) type {
 
         /// to deinit channels created with `create`
         pub fn destroy(channel: *Self) void {
+            const allocator = channel.allocator; // read allocator from channel before freeing it
             channel.deinit();
-            channel.allocator.destroy(channel);
+            allocator.destroy(channel);
         }
 
         pub fn close(channel: *Self) void {
             channel.closed.store(true, .monotonic);
-            channel.condition.broadcast();
         }
 
         pub fn send(channel: *Self, value: T) !void {
             if (channel.closed.load(.monotonic)) {
                 return error.ChannelClosed;
             }
+
+            const send_hook = channel.send_hook;
+            if (send_hook) |hook| {
+                if (!hook.before_send(hook, channel, value)) {
```

Comment:
i might be confused but shouldnt it just be `hook.before_send(channel, value)`

Reply:
The function is able to return a bool indicating whether the send is actually necessary. The boolean indicates whether to send it. When it returns false, that means "we already handled it, no need to send it." That way you don't get redundant processing of the message.

Reply:
https://github.com/Syndica/sig/pull/486/commits/767607c3c46beb88cf4533d9bbabba71b9b3e634 (removed `beforeSend` for now, reasoning: https://github.com/Syndica/sig/pull/486#discussion_r1917161745)